File: 2024/kotlin/src/day01/Day01.java
package day01;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.IntStream;

import static utils.UtilsKt.readInput;

public class Day01 {
    static final String WHITE_SPACE = "   ";
    static final String testInput = readInput("test");
    static final String input = readInput("input");

    public static void main(String[] args) {
        System.out.printf("Part 1 test: %d", part1(testInput));
        System.out.printf("Part 1 input: %d", part1(input));
    }

    static int part1(String input) {
        List<Integer> left = new ArrayList<>();
        List<Integer> right = new ArrayList<>();

        input.lines().map(l -> l.split(WHITE_SPACE)).forEach(parts -> {
            left.add(Integer.parseInt(parts[0]));
            right.add(Integer.parseInt(parts[1]));
        });

        left.sort(null);
        right.sort(null);

        return IntStream.range(0, left.size()).map(i -> Math.abs(left.get(i) - right.get(i))).sum();
    }
}

File: 2024/kotlin/src/day01/Day01.kt
package day01

import utils.parseInput
import utils.println
import kotlin.math.abs


fun main() {
    val part1Expected = 11
    val part2Expected = 31

    fun part1(input: List<String>): Int {
        val leftNumbers = input.map {
            it.split("   ").first().toInt()
        }.sorted()
        val rightNumbers = input.map {
            it.split("   ").last().toInt()
        }.sorted()

        val pairs = leftNumbers.zip(rightNumbers)

        return pairs.sumOf {
            abs(it.first - it.second)
        }
    }

    fun part2(input: List<String>): Int {
        val leftNumbers = input.map {
            it.split("   ").first().toInt()
        }
        val rightNumbers = input.map {
            it.split("   ").last().toInt()
        }

        return leftNumbers.sumOf {
            val appearances = rightNumbers.count { right -> it == right }
            it * appearances
        }
    }

    val testInput = parseInput("example1")
    check(part1(testInput) == part1Expected)
    check(part2(testInput) == part2Expected)

    val input = parseInput("input")
    part1(input).println("part1")
    part2(input).println("part2")
}
